Migrate TS SignalR client to ESLint

joaopgrassi opened the pull request:
As detailed in #30440, TSLint is deprecated. This PR migrates the SignalR client from TSLint to ESLint. 

Resume of the changes:

- Installed [Typescript ESLint](https://github.com/typescript-eslint/typescript-eslint)
- Used the recommended rules from them (which also includes the base ones from ESLint)
- Found the rules that were enabled by default in TSLint and not in ESLint/TS ESlint and added them to the rules
- "Converted" the existing rules to ESLint
- Some errors/warnings appeared, so minor fixes had to be applied
- Removed all traces of TSLint from code

Fixes #30440

joaopgrassi committed:
Initial fix due to ESlint

joaopgrassi committed:
WIP: Initial setup with ESlint

joaopgrassi committed:
More fixes after enabling eslint

joaopgrassi committed:
WIP: Adding rules

joaopgrassi committed:
WIP: Adding more rules
Should have now everything that was ON by default in TSLint but was OFF by default in ESLint/typescript-eslint

joaopgrassi committed:
Use ESLint for pretest task
Pending to add to msgpack project still

joaopgrassi committed:
Fix ESLint warnings/errors in tests

joaopgrassi committed:
Use ESLint for msgpack project

joaopgrassi committed:
Remove all left over from TSLint

joaopgrassi commented on src/SignalR/clients/ts/FunctionalTests/ts/ConnectionTests.ts:
```diff
@@ -1,8 +1,8 @@
-// This code uses a lot of `.then` instead of `await` and TSLint doesn't like it.
-// tslint:disable:no-floating-promises
+// This code uses a lot of `.then` instead of `await` and ESLint doesn't like it.
+/* eslint-disable @typescript-eslint/no-floating-promises */
```
It complains because of the `connection.stop()` we have, for example, at line 32. 

```
connection.onreceive = (data: any) => {
	if (data === message) {
		await connection.stop();
	}
};
```

joaopgrassi replied:
Done.

joaopgrassi committed:
PR suggestions
Removed no-empty-function rule and fixed tests missing `await`

joaopgrassi commented on src/SignalR/clients/ts/FunctionalTests/ts/HubConnectionTests.ts:
```diff
@@ -167,6 +167,7 @@ describe("hubConnection", () => {
 
                 await hubConnection.start();
                 const subscription = hubConnection.stream<string>("InfiniteStream").subscribe({
+                    /* eslint-disable @typescript-eslint/no-empty-function */
```
Done.

joaopgrassi commented on src/SignalR/clients/ts/FunctionalTests/ts/HubConnectionTests.ts:
```diff
@@ -1,8 +1,8 @@
-// This code uses a lot of `.then` instead of `await` and TSLint doesn't like it.
-// tslint:disable:no-floating-promises
+// This code uses a lot of `.then` instead of `await` and ESLint doesn't like it.
+/* eslint-disable @typescript-eslint/no-floating-promises */
```
Done

joaopgrassi commented on src/SignalR/clients/ts/FunctionalTests/ts/Utils.ts:
```diff
@@ -1,8 +1,9 @@
-export function getParameterByName(name: string) {
+export function getParameterByName(name: string): string | null {
     const url = window.location.href;
+    // eslint-disable-next-line no-useless-escape
```
It actually complains with: `Unnecessary escape character: \[`. I guess both `\[` and `\]` are unnecessary?

joaopgrassi replied:
haha I had the same thought :sweat_smile:

joaopgrassi commented on src/SignalR/clients/ts/signalr/src/LongPollingTransport.ts:
```diff
@@ -111,13 +111,13 @@ export class LongPollingTransport implements ITransport {
         if (token) {
-            // tslint:disable-next-line:no-string-literal
+            // eslint-disable-next-line @typescript-eslint/dot-notation
             request.headers["Authorization"] = `Bearer ${token}`;
```
Yeah I was not sure about these. We could change to use dot notation, like in this case it would work fine. In other places, we might need to add the type, like `const headers: MessageHeaders = {};`, so we can do `headers.Cookie = `${cookies}`;`. 

Another approach is have these header names as constants somewhere. Then it would not complain: 
`headers[COOKIE_HEADER] = `${cookies}`;` like so.

joaopgrassi replied:
Introduced a `HeaderNames` file containing only cookie and authorization. I guess we can add more as the need surfaces. Let me know what you think :)

joaopgrassi commented on src/SignalR/clients/ts/signalr/tests/HttpConnection.test.ts:
```diff
@@ -235,10 +235,12 @@ describe("HttpConnection", () => {
         "either the endpoint may not be a SignalR endpoint, the connection ID is not present on the server, or there is a proxy blocking WebSockets. If you have multiple servers check that sticky sessions are enabled.",
         "Failed to start the connection: Error: Unable to connect to the server with any of the available transports. WebSockets failed: " +
         "Error: WebSocket failed to connect. The connection could not be found on the server, either the endpoint may not be a SignalR endpoint, the connection ID is not present on the server, or there is a proxy blocking WebSockets. If you have multiple servers check that sticky sessions are enabled. ServerSentEvents failed: Error: 'ServerSentEvents' is disabled by the client. LongPolling failed ...
+        /* eslint-disable max-len */
```
This rule was enabled before in our TSLint: `"max-line-length": { "options": [300] },`. I kept the same length, but for some reason it complained.

joaopgrassi replied:
Ah you are right, forgot to enable it again. Fixed.

joaopgrassi committed:
Removed leading space

joaopgrassi commented on src/SignalR/clients/ts/signalr/src/Utils.ts:
```diff
@@ -233,6 +233,7 @@ export function constructUserAgent(version: string, os: string, runtime: string,
 }
 
+// eslint-disable-next-line spaced-comment
  /*#__PURE__*/ function getOsName(): string {
```
Done. Also removed for the other below :)

joaopgrassi committed:
Enable name-convention rule
Reverted places were the rule should be disabled since it's used in tests.

joaopgrassi commented on src/SignalR/clients/ts/signalr/src/HubConnection.ts:
```diff
@@ -643,11 +648,10 @@ export class HubConnection {
         }
     }
 
-    // tslint:disable-next-line:naming-convention
```
I pushed a new commit with the `naming-convetion` enabled requiring `_` for member-like. I also restored this disable and also in `Errors.ts`.

BrennanConroy commented:
Looks like this needs a rebase on master. The clients/ts/common/yarn.lock file got updated.

joaopgrassi committed:
Re-enable max-len

joaopgrassi commented on src/SignalR/clients/ts/signalr/src/Utils.ts:
```diff
@@ -233,7 +233,8 @@ export function constructUserAgent(version: string, os: string, runtime: string,
 }
 
- /*#__PURE__*/ function getOsName(): string {
+// eslint-disable-next-line spaced-comment
```
Ah I see what you meant now. No, this rule enforces that comments start with a space. So it complains and it wants it to be `/* #__PURE__*/, but I'm not sure if it the space should be there. I'm not familiar with this, so I just disabled it here.

joaopgrassi committed:
WIP: Adding rules
Adding rules that were enabled by default in TSLint and are not enabled by default in ESLint

joaopgrassi committed:
WIP: Adding rules
Adding rules that were enabled by default in TSLint and are not enabled by default in ESLint

joaopgrassi committed:
Rebase main and fix yarn.lock conflicts
Took lock file from master and re-added the ESLint packages.

joaopgrassi commented:
Did a rebase with the latest main. I just dropped my changes, did the rebase then re-added the packages, since figuring out what changed was hard.

joaopgrassi committed:
Introduce constants for HTTP header names
In order to fix the dot-notation ESLint rule and avoid string repetition.

joaopgrassi committed:
Fix import for header names file

joaopgrassi commented on src/SignalR/clients/ts/FunctionalTests/ts/HubConnectionTests.ts:
```diff
@@ -1056,6 +1053,7 @@ describe("hubConnection", () => {
     function getJwtToken(url: string): Promise<string> {
         return new Promise((resolve, reject) => {
             const httpClient = new DefaultHttpClient(NullLogger.instance);
+            // eslint-disable-next-line @typescript-eslint/no-floating-promises
```
You mean convert the `getJwtToken` func to async and get rid of the inner Promise or make the inner executor async? If the latter, it will run into this: https://eslint.org/docs/rules/no-async-promise-executor. What it's probably better is this:

```
async function getJwtToken(url: string): Promise<string> {
        const httpClient = new DefaultHttpClient(NullLogger.instance);
        const response = await httpClient.get(url);

        if (response.statusCode >= 200 && response.statusCode < 300) {
            return response.content as string;
        } else {
            throw new Error(response.statusText);
        }
    }
```

Just not sure about the throw. What you think?
The functional tests are not running on my box bcs I don't have the preview sdk installed.. seems that on debian it's not so easy to install all so I just didn't do it hehe.

BrennanConroy commented on src/SignalR/clients/ts/signalr/src/HeaderNames.ts:
```diff
@@ -0,0 +1,4 @@
+export abstract class HeaderNames {
```
```suggestion
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.

export abstract class HeaderNames {
```

BrennanConroy committed:
Update src/SignalR/clients/ts/signalr/src/HeaderNames.ts

BrennanConroy reviewed (approved):
This is great!

BrennanConroy commented:
Thanks again @joaopgrassi !
